source/grid/grid_tools.cc

```diff
@@ -6452,6 +6458,479 @@ namespace GridTools
+                result.recv_components.size(), // number of point
+                numbers::invalid_unsigned_int);
+            }
+        }
+
+      // since empty quadratures might be present we have to set the number
+      // of searched points after inserting the point indices into
+      // recv_components
+      result.n_searched_points = result.recv_components.size();
+
+#ifdef DEAL_II_WITH_MPI
+      // send_ranks, counter, and indices_of_rank is only needed if
+      // consistent_numbering_of_sender_and_receiver==true
+      std::set<unsigned int>                            send_ranks;
+      std::map<unsigned int, unsigned int>              counter;
+      std::map<unsigned int, std::vector<unsigned int>> indices_of_rank;
+      if (consistent_numbering_of_sender_and_receiver)
+        {
+          for (const auto &sc : send_components)
+            send_ranks.insert(std::get<1>(sc));
+
+          for (const auto rank : send_ranks)
+            counter[rank] = 0;
+
+          // indices assigned at recv side needed to fill send_components
+          indices_of_rank = communicate_indices(result.recv_components,
+                                                tria.get_communicator());
+        }
+#else
+      (void)consistent_numbering_of_sender_and_receiver;
+#endif
+
+      for (const auto &send_component : send_components)
+        {
+          CellIterator cell = {&tria,
+                               std::get<0>(send_component).first,
+                               std::get<0>(send_component).second};
+
+          const Quadrature<spacedim> &quad =
+            quadrature.compute_affine_transformation(
+              std::get<3>(send_component));
+
+          const auto rank = std::get<1>(send_component);
+
+          for (unsigned int q = 0; q < quad.size(); ++q)
+            {
+              // the fifth component of result.send_components is filled
+              // during sorting the data and initializing the CRS structures
+              result.send_components.emplace_back(std::make_tuple(
+                std::get<0>(send_component),
+                rank,
+#ifdef DEAL_II_WITH_MPI
+                !consistent_numbering_of_sender_and_receiver ?
+                  result.send_components.size() :
+                  indices_of_rank.at(rank)[counter.at(rank)],
+#else
+                result.send_components.size(),
+#endif
+                mapping.transform_real_to_unit_cell(cell, quad.point(q)),
+                quad.point(q),
+                numbers::invalid_unsigned_int));
+
+#ifdef DEAL_II_WITH_MPI
+              if (consistent_numbering_of_sender_and_receiver)
+                ++counter[rank];
+#endif
+            }
+        }
+
+      result.finalize_setup();
+
+      return result;
+    }
+
+
+
+    template <int structdim, int spacedim>
+    std::map<unsigned int, std::vector<unsigned int>>
+    DistributedComputeIntersectionLocationsInternal<structdim, spacedim>::
+      communicate_indices(
```

Comment:
I am not particular fond of the implementation of this function; I think an implementation without pack/unpack (that works directly on a single vector) would be much nicer. But can do that change in a follow-up PR.

Reply:
I added this as a follow-up to the issue :)

Reply:
@peterrum Do we really have to address this? `communicate_indices()` is only left, because it is helpful for debugging anyway... It is not needed in productive code.

Reply:
Let's keep it like this.